File: test-client.js
// Simple Node.js test client for the 2D Datamatrix Detection Service
// Usage: node test-client.js <path-to-image>

const fs = require('fs');
const axios = require('axios');
const FormData = require('form-data');
const path = require('path');

// Configuration
const API_URL = 'http://localhost:5001';

// Check command line arguments
if (process.argv.length < 3) {
  console.error('Usage: node test-client.js <path-to-image>');
  process.exit(1);
}

const imagePath = process.argv[2];

// Ensure the image file exists
if (!fs.existsSync(imagePath)) {
  console.error(`Error: File ${imagePath} does not exist`);
  process.exit(1);
}

// Test file upload endpoint
async function testFileUpload() {
  console.log(`Testing file upload with ${imagePath}`);
  
  try {
    // Create form data with the image file
    const formData = new FormData();
    formData.append('image', fs.createReadStream(imagePath));
    
    // Send request to the detection service
    const response = await axios.post(`${API_URL}/detect`, formData, {
      headers: {
        ...formData.getHeaders(),
      },
    });
    
    // Process the response
    const result = response.data;
    console.log(`Success! Found ${result.count} codes`);
    console.log(`Image URL: ${result.image_url}`);
    
    // Print detected codes
    result.detected_codes.forEach((code, i) => {
      console.log(`Code ${i+1}:`);
      console.log(`  Method: ${code.method}`);
      console.log(`  Data: ${code.data}`);
      console.log(`  Type: ${code.type}`);
      console.log(`  Position:`, code.position);
      console.log();
    });
    
    return result;
  } catch (error) {
    console.error('Error:', error.response?.status || error.message);
    console.error(error.response?.data || error);
    return null;
  }
}

// Test base64 upload endpoint
async function testBase64Upload() {
  console.log(`Testing base64 upload with ${imagePath}`);
  
  try {
    // Read the image and encode it as base64
    const imageBuffer = fs.readFileSync(imagePath);
    const base64Image = imageBuffer.toString('base64');
    
    // Prepare the JSON payload
    const payload = {
      image: base64Image,
      include_image: true  // Include the processed image in the response
    };
    
    // Send request to the detection service
    const response = await axios.post(`${API_URL}/detect_base64`, payload, {
      headers: {
        'Content-Type': 'application/json',
      },
    });
    
    // Process the response
    const result = response.data;
    console.log(`Success! Found ${result.count} codes`);
    
    // Print detected codes
    result.detected_codes.forEach((code, i) => {
      console.log(`Code ${i+1}:`);
      console.log(`  Method: ${code.method}`);
      console.log(`  Data: ${code.data}`);
      console.log(`  Type: ${code.type}`);
      console.log(`  Position:`, code.position);
      console.log();
    });
    
    // Save the returned image if it's included
    if (result.image) {
      // Extract the base64 part
      const imgData = result.image.split(',')[1];
      const imgBuffer = Buffer.from(imgData, 'base64');
      
      // Save to file
      const outputPath = 'received_image.jpg';
      fs.writeFileSync(outputPath, imgBuffer);
      console.log(`Saved processed image to ${outputPath}`);
    }
    
    return result;
  } catch (error) {
    console.error('Error:', error.response?.status || error.message);
    console.error(error.response?.data || error);
    return null;
  }
}

// Run the tests
async function runTests() {
  console.log('=== Testing 2D Datamatrix Detection Service ===\n');
  
  // Test file upload
  await testFileUpload();
  
  console.log('\n' + '-'.repeat(50) + '\n');
  
  // Test base64 upload
  await testBase64Upload();
}

runTests().catch(console.error);

File: test_client.py
import requests
import base64
import json
import sys
import os

def test_file_upload(image_path):
    """Test the /detect endpoint with file upload"""
    print(f"Testing file upload with {image_path}")
    
    url = "http://localhost:5001/detect"
    
    # Check if file exists
    if not os.path.exists(image_path):
        print(f"Error: File {image_path} does not exist")
        return
    
    # Prepare the file for upload
    files = {'image': open(image_path, 'rb')}
    
    # Make the request
    response = requests.post(url, files=files)
    
    # Print the response
    if response.status_code == 200:
        result = response.json()
        print(f"Success! Found {result['count']} codes")
        print(f"Image URL: {result['image_url']}")
        
        # Print detected codes
        for i, code in enumerate(result['detected_codes']):
            print(f"Code {i+1}:")
            print(f"  Method: {code['method']}")
            print(f"  Data: {code['data']}")
            print(f"  Type: {code['type']}")
            print(f"  Position: {code['position']}")
            print()
    else:
        print(f"Error: {response.status_code}")
        print(response.text)

def test_base64_upload(image_path):
    """Test the /detect_base64 endpoint with base64 encoded image"""
    print(f"Testing base64 upload with {image_path}")
    
    url = "http://localhost:5001/detect_base64"
    
    # Check if file exists
    if not os.path.exists(image_path):
        print(f"Error: File {image_path} does not exist")
        return
    
    # Read the image and encode it as base64
    with open(image_path, 'rb') as image_file:
        encoded_string = base64.b64encode(image_file.read()).decode('utf-8')
    
    # Prepare the JSON payload
    payload = {
        'image': encoded_string,
        'include_image': True  # Include the processed image in the response
    }
    
    # Make the request
    headers = {'Content-Type': 'application/json'}
    response = requests.post(url, data=json.dumps(payload), headers=headers)
    
    # Print the response
    if response.status_code == 200:
        result = response.json()
        print(f"Success! Found {result['count']} codes")
        
        # Print detected codes
        for i, code in enumerate(result['detected_codes']):
            print(f"Code {i+1}:")
            print(f"  Method: {code['method']}")
            print(f"  Data: {code['data']}")
            print(f"  Type: {code['type']}")
            print(f"  Position: {code['position']}")
            print()
        
        # Save the returned image if it's included
        if 'image' in result:
            # Extract the base64 part
            img_data = result['image'].split(',')[1]
            img_bytes = base64.b64decode(img_data)
            
            # Save to file
            output_path = "received_image.jpg"
            with open(output_path, 'wb') as f:
                f.write(img_bytes)
            print(f"Saved processed image to {output_path}")
    else:
        print(f"Error: {response.status_code}")
        print(response.text)

if __name__ == "__main__":
    # Check if image path is provided
    if len(sys.argv) < 2:
        print("Usage: python test_client.py <image_path>")
        sys.exit(1)
    
    image_path = sys.argv[1]
    
    # Test both endpoints
    test_file_upload(image_path)
    print("\n" + "-"*50 + "\n")
    test_base64_upload(image_path)